### PR TITLE
Frontend intermediate representation

### PR DESCRIPTION
**Pull request by fabfors**
Added an intermediate representation to the Zedex BNFC-generated AST. Also added conversion and reversion functions between these two AST's.

### CODE REVIEW

File: src/Frontend/IAST/Abs.hs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  deriving (Eq, Show, Read)
+  
+newtype Program = Progr [Toplevel]
+  deriving (Eq, Show, Read) 
+
+data Basis
+    = Z
+    | X
+    | None
+  deriving (Eq, Show, Read)
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Abs.hs:18]: newline

File: src/Frontend/IAST/Abs.hs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import Data.Complex (Complex)
+
+data Id = Id
+    { pos :: (Int, Int)
+    , name :: String 
+    }
+  deriving (Eq, Show, Read)
+  
+newtype Program = Progr [Toplevel]
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Abs.hs:11]: No need for a newtype here, since it conveys no information.
```suggestion
type Program = [Toplevel]
```

**Review by lambducas — COMMENTED** [review]
Looks good. Fancy might I add.

File: src/Frontend/IAST/Convert.hs [hunk 3]
```diff
@@ -0,0 +1,89 @@
+module Frontend.IAST.Convert where
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Convert.hs:1]: List module exports.

File: src/Frontend/IAST/Abs.hs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    | None
+  deriving (Eq, Show, Read) 
+data Toplevel
+    = ToplF Id [Id] Expr
+    | Topl Basis (Complex Double) [Id] Expr
+  deriving (Eq, Show, Read)
+  
+data Expr
+    = Var Id
+    | Tup Expr [Expr]
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Abs.hs:26]: Convert this to a single expr-list. The reason for it being two expressions in the parser is the requirement for tuples to include at least 2 elements.
```suggestion
    | Tup [Expr]
```

Comment (NicklasBoto) [reply src/Frontend/IAST/Abs.hs:26]: > Or even use a non-empty list here, or a sized list.

File: src/Frontend/IAST/Convert.hs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+convertExpr :: ZX.Expr -> IAST.Expr
+convertExpr (ZX.EVar var) = IAST.Var (convertId var)
+convertExpr (ZX.ETup e es) = IAST.Tup (convertExpr e) (map convertExpr es)
+convertExpr ZX.EUnit = IAST.Unit
+convertExpr (ZX.EApp e1 e2) = IAST.App (convertExpr e1) (convertExpr e2)
+convertExpr (ZX.EComp e1 e2) = IAST.Comp (convertExpr e1) (convertExpr e2)
+convertExpr (ZX.EHad e) = IAST.Had (convertExpr e)
+convertExpr (ZX.EAbs ids e) = IAST.Abs IAST.None 0 (convertIds ids) (convertExpr e)
+convertExpr (ZX.EXAbs c ids e) = IAST.Abs IAST.X (complexConvert c) (convertIds ids) (convertExpr e)
+convertExpr (ZX.EZAbs c ids e) = IAST.Abs IAST.Z (complexConvert c) (convertIds ids) (convertExpr e)
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Convert.hs:25]: Re-format

File: src/Frontend/IAST/Convert.hs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+convert (ZX.Progr topls) = IAST.Progr $ map convertTopl topls
+
+convertTopl :: ZX.Toplevel -> IAST.Toplevel
+convertTopl (ZX.ToplF name ids e) = IAST.ToplF (convertId name) (convertIds ids) (convertExpr e)
+convertTopl (ZX.ToplX c ids e)    = IAST.Topl IAST.X (complexConvert c) (convertIds ids) (convertExpr e)
+convertTopl (ZX.ToplZ c ids e)    = IAST.Topl IAST.Z (complexConvert c) (convertIds ids) (convertExpr e)
+
+convertExpr :: ZX.Expr -> IAST.Expr
+convertExpr (ZX.EVar var) = IAST.Var (convertId var)
+convertExpr (ZX.ETup e es) = IAST.Tup (convertExpr e) (map convertExpr es)
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Convert.hs:18]:
```suggestion
convertExpr (ZX.ETup e es) = IAST.Tup (map convertExpr (e:es))
```

File: src/Frontend/IAST/Convert.hs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+convertExpr (ZX.EZAbs c ids e) = IAST.Abs IAST.Z (complexConvert c) (convertIds ids) (convertExpr e)
+
+complexConvert :: ZX.Complex -> Complex Double
+complexConvert (ZX.CComp (ZX.Scalar s1) (ZX.Scalar s2)) = read s1 :+ read s2
+complexConvert (ZX.CComn (ZX.Scalar s1) (ZX.Scalar s2)) = read s1 :+ negate (read s2)
+complexConvert ZX.CPi = pi :+ 0
+complexConvert ZX.CE = exp 1 :+ 0
+complexConvert (ZX.CExp c1 c2) = complexConvert c1**complexConvert c2
+complexConvert (ZX.CDiv c1 c2) = complexConvert c1/complexConvert c2
+complexConvert (ZX.CMul c1 c2) = complexConvert c1*complexConvert c2
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Convert.hs:34]: Formatting, add spaces between binary operators

File: src/Frontend/IAST/Convert.hs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+          ids = [id1, id2]
+
+testIASTPrg :: IAST.Program
+testIASTPrg = IAST.Progr [topl]
+    where topl = IAST.Topl IAST.Z (0 :+ 0) ids expr
+          expr = IAST.App (IAST.Abs IAST.Z (0 :+ 0) [id1] (IAST.Had (IAST.Var id1)))
+                (IAST.Var id2)
+          id1 = IAST.Id (0,0) "id1"
+          id2 = IAST.Id (0,1) "id2"
+          ids = [id1,id2]
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Convert.hs:89]: Don't export these

**Review by NicklasBoto — CHANGES_REQUESTED** [review]
Looks nice! Please look at the suggestions I wrote.

File: src/Frontend/IAST/Convert.hs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+revertTopl :: IAST.Toplevel -> ZX.Toplevel
+revertTopl (IAST.ToplF name ids e)       = ZX.ToplF (revertId name) (revertIds ids) (revertExpr e)
+revertTopl (IAST.Topl IAST.Z c ids e)    = ZX.ToplZ (revertComplex c) (revertIds ids) (revertExpr e)
+revertTopl (IAST.Topl IAST.X c ids e)    = ZX.ToplX (revertComplex c) (revertIds ids) (revertExpr e)
+revertTopl (IAST.Topl IAST.None c ids e) = ZX.ToplF ((revertId . head) ids) (revertIds ids) (revertExpr e)
+
+revertExpr :: IAST.Expr -> ZX.Expr
+revertExpr (IAST.Var var)            = ZX.EVar (revertId var)
+revertExpr (IAST.Tup (e:es))         = ZX.ETup (revertExpr e) (map revertExpr es)
```
Comment (NicklasBoto) [inline src/Frontend/IAST/Convert.hs:53]: This makes the function partial for tuple constructors. Do we think it is necessary to catch this? With what though..
```suggestion
revertExpr (IAST.Tup (e:es)) = ZX.ETup (revertExpr e) (map revertExpr es)
revertExpr (IAST.Tup [])     = ZX.Unit
```

**Review by NicklasBoto — CHANGES_REQUESTED** [review]
Only one comment now!

**Review by NicklasBoto — APPROVED** [review]
Nice!

**Pull request merged**